EKF/control.cpp

```diff
@@ -580,16 +581,28 @@ void Ekf::controlGpsFusion()
 				_time_last_on_ground_us = _time_last_imu;
 			}
 
-			const bool recent_takeoff_nav_failure = _control_status.flags.in_air &&
+			bool recent_takeoff_nav_failure = _control_status.flags.in_air &&
 								!isTimedOut(_time_last_on_ground_us, 30000000) &&
 								isTimedOut(_time_last_hor_vel_fuse, _params.EKFGSF_reset_delay) &&
 								(_time_last_hor_vel_fuse > _time_last_on_ground_us);
 
-			const bool inflight_nav_failure = _control_status.flags.in_air &&
+			bool inflight_nav_failure = _control_status.flags.in_air &&
 							  do_vel_pos_reset &&
 							  (_time_last_hor_vel_fuse > _time_last_on_ground_us) &&
 							  (_time_last_hor_pos_fuse > _time_last_on_ground_us);
 
+			if ((recent_takeoff_nav_failure || inflight_nav_failure) && _time_last_hor_vel_fuse > 0) {
+				// Do sanity check to see if the innovation failures is likely caused by a yaw angle error
+				// A 180 deg yaw error would (in the absence of other errors) generate a velocity innovation
+				// magnitude of twice the measured speed so allow up to 3 times to allow for additional errors.
+				const float hvel_obs_sq = sq(_last_vel_obs(0)) + sq(_last_vel_obs(1));
+				const float hvel_innov_sq = sq(_last_fail_hvel_innov(0)) + sq(_last_fail_hvel_innov(1));
+				if (hvel_innov_sq > 9.0f * hvel_obs_sq) {
+					recent_takeoff_nav_failure = false;
+					inflight_nav_failure = false;
```

Comment:
As described in the comment, this check is the answer to "is the nav issues due to a yaw error?", so IMO, instead of changing `recent_takeoff_nav_failure` and `inflight_nav_failure`, we should rather add a new boolean `is_yaw_failure` and replace the check below:
```c++
const bool do_yaw_vel_pos_reset = (_do_ekfgsf_yaw_reset || recent_takeoff_nav_failure || inflight_nav_failure) && ...
```
So the meaning of the variables isn't changed.
by 
```c++
const bool do_yaw_vel_pos_reset = (_do_ekfgsf_yaw_reset || is_yaw_failure) && ...
```

Reply:
Done